File: services/fae/go/fae/report.go
package fae

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/json"
)

// GenerateReport generates a deterministic JSON document with a signature.
func GenerateReport(payload any, secret string) ([]byte, error) {
	canonical, err := canonicalJSON(map[string]any{"payload": payload})
	if err != nil {
		return nil, err
	}
	sig := hmac.New(sha256.New, []byte(secret))
	sig.Write(canonical)
	signature := sig.Sum(nil)
	report, err := canonicalJSON(map[string]any{
		"payload":   payload,
		"signature": encodeHex(signature),
	})
	if err != nil {
		return nil, err
	}
	return report, nil
}

// VerifyReport verifies the signature against the secret.
func VerifyReport(report []byte, secret string) (bool, error) {
	var document map[string]any
	if err := json.Unmarshal(report, &document); err != nil {
		return false, err
	}
	signature, _ := document["signature"].(string)
	payload := map[string]any{"payload": document["payload"]}
	canonical, err := canonicalJSON(payload)
	if err != nil {
		return false, err
	}
	sig := hmac.New(sha256.New, []byte(secret))
	sig.Write(canonical)
	expected := encodeHex(sig.Sum(nil))
	return hmac.Equal([]byte(signature), []byte(expected)), nil
}

func canonicalJSON(payload any) ([]byte, error) {
	return json.Marshal(payload)
}

func encodeHex(data []byte) string {
	const hexdigits = "0123456789abcdef"
	dst := make([]byte, len(data)*2)
	for i, b := range data {
		dst[i*2] = hexdigits[b>>4]
		dst[i*2+1] = hexdigits[b&0x0f]
	}
	return string(dst)
}

File: services/fae/python/fae/report.py
"""Signed evaluation report generation."""

from __future__ import annotations

import hmac
import json
from dataclasses import asdict, is_dataclass
from hashlib import sha256
from typing import Any, Mapping, Sequence


def _serialize(obj: Any) -> Any:
    if is_dataclass(obj):
        return {k: _serialize(v) for k, v in asdict(obj).items()}
    if isinstance(obj, Mapping):
        return {str(k): _serialize(v) for k, v in sorted(obj.items())}
    if isinstance(obj, Sequence) and not isinstance(obj, (str, bytes, bytearray)):
        return [_serialize(v) for v in obj]
    return obj


def _canonical_json(payload: Mapping[str, Any]) -> bytes:
    return json.dumps(payload, sort_keys=True, separators=(",", ":")).encode("utf-8")


def generate_report(payload: Mapping[str, Any], secret: str) -> bytes:
    serialized = _serialize(payload)
    canonical = _canonical_json({"payload": serialized})
    signature = hmac.new(secret.encode("utf-8"), canonical, sha256).hexdigest()
    return _canonical_json({"payload": serialized, "signature": signature})


def verify_report(report_bytes: bytes, secret: str) -> bool:
    document = json.loads(report_bytes.decode("utf-8"))
    signature = document.get("signature", "")
    payload = {"payload": document.get("payload")}
    canonical = _canonical_json(payload)
    expected = hmac.new(secret.encode("utf-8"), canonical, sha256).hexdigest()
    return hmac.compare_digest(signature, expected)

